core/shared/platform/zephyr/zephyr_file.c

```diff
@@ -0,0 +1,967 @@
+/*
+ * Copyright (C) 2024 Grenoble INP - ESISAR.  All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#include "platform_api_extension.h"
+#include "libc_errno.h"
+
+#include <string.h>
+#include <stdlib.h>
+
+#include <zephyr/fs/fs.h>
+#include <zephyr/fs/fs_interface.h>
+#include <zephyr/fs/fs_sys.h>
+#include <zephyr/fs/littlefs.h>
+
+/* Notes:
+ * This is the implementation of a POSIX-like file system interface for Zephyr.
+ * To manage our file descriptors, we created a struct `zephyr_fs_desc` that
+ * represent a zephyr file descriptor and hold useful informations.
+ * We also created a file descriptor table to keep track of all the file
+ * descriptors.
+ * To pass the file descriptor reference to the higher level abstraction, we
+ * pass the index of the fd table to an `os_file_handle` struct.
+ * Then in the WASI implementation layer we can retrieve the file descriptor
+ * reference.
+ * We also fake the stdin, stdout and stderr file descriptors.
+ * We redirect the write operation on stdin, stdout and stderr to `os_printf`.
+ * We do not handle write on stdin and read on stdin, stdout and stderr.
+ */
+
+// No OS API wrapper (Zephyr):
+// file:
+//     off_t fs_tell(struct fs_file_t *zfp)
+// file system:
+//     int fs_mount(struct fs_mount_t *mp)
+//     int fs_unmount(struct fs_mount_t *mp
+//     int fs_readmount(int *index, const char **name)
+//     int fs_statvfs(const char *path, struct fs_statvfs *stat)
+//     int fs_mkfs(int fs_type, uintptr_t dev_id, void *cfg, int flags)
+//     int fs_register(int type, const struct fs_file_system_t *fs)
+//     int fs_unregister(int type, const struct fs_file_system_t *fs)
+
+// We will take the maximum number of open files
+// from the Zephyr POSIX configuration
+#define CONFIG_WASI_MAX_OPEN_FILES 16
+
+// Macro to retrieve a file system descriptor and check it's validity.
+#define GET_FILE_SYSTEM_DESCRIPTOR(fd, ptr) \
+    do {                                    \
+        ptr = &desc_array[(int)fd];         \
+        if (!ptr)                           \
+            return __WASI_EBADF;            \
+    } while (0)
+
+// Array to keep track of file system descriptors.
+static struct zephyr_fs_desc desc_array[CONFIG_WASI_MAX_OPEN_FILES];
+
+// mutex to protect the file descriptor array
+K_MUTEX_DEFINE(desc_array_mutex);
```

Comment:
The mutex is only used in zephyr_fs_alloc_obj to protect desc_array, but desc_array is used in many places when GET_FILE_SYSTEM_DESCRIPTOR is called, seems we should also use the mutex to protect them?

Reply:
I locked the access to the array in the GET_FILE_SYSTEM_DESCRIPTOR macro.